hw07.c

```diff
@@ -0,0 +1,246 @@
+int tgaheader_width( const TGAheader * self ) {
+    int width = 0;
+
+    memcpy( &width, self->width, 2 );
+    printf("width: %d\n",width);
+
+    return width;
+}
+
+int tgaheader_height( const TGAheader * self ) {
+    int height = 0;
+
+    memcpy( &height, self->height, 2 );
+    printf("height: %d\n",height);
+
+    return height;
+
+}
+
+int tgaheader_channels( const TGAheader * self ) {
+    return self->depth / 8;
+}
+
+//zjistovani hodnot z headru pres fci - asi prehlednejsi
+int tga_width( const TGAImage * self ) {
+    return tgaheader_width( &self->header );
+}
+
+int tga_height( const TGAImage * self ) {
+    return tgaheader_height( &self->header );
+}
+
+int tga_channels( const TGAImage * self ) {
+    return  tgaheader_channels( &self->header );
+}
+
+
+//prevodova fce z 2d souradnic do indexu 1D odpovidajiciho serioveho elementu
+int to1D(int row, int col, int cols){
+    return col + row * cols;
+}
+
+
+//cokoliv zavola fci se souradnicema x,y bude prevedeno z 2d pole do odpovidajiciho indexu 1d pole jiz nainacializovaneho v *data
+//prehodim x, y kdyz pocatek je 0,0 v dolnim levem rohy pokud prehodim x z y tak he to jakobych ho otozil o 90/ prohodil sloupce/radky
+void set_pixel(TGAImage *self, int y, int x, RGBA *pix){
+    self->data[to1D(x,y, tga_height(self))].alpha = pix->alpha;
+    self->data[to1D(x,y, tga_height(self))].blue = pix->blue;
+    self->data[to1D(x,y, tga_height(self))].green = pix->green;
+    self->data[to1D(x,y, tga_height(self))].red = pix->red;
+}
+
+void draw_bg(TGAImage *self, RGBA *bg){
+    for (int i = 0; i < tga_height(self); i++)
+    {
+        for (int j = 0; j < tga_width(self); j++)
+        {
+            set_pixel(self, i ,j , bg);
+            printf("bg: i:%d j:%d\n",i ,j);
+        }
+
+    }
+
+}
+
+//pocita indexy nalezejici vykresleni znaku  ty posila do fce ktere zapisuje ty hodnotu do pole clena strukty *data
+void draw_one(TGAImage *self, int dx, int dy, RGBA *pix){
+    for (int i = 0 + dx; i < 195 + dx; i++)
+    {
+        for (int j = 0 + dx; j < 169 + dy; j++)
+        {
+            if (i > 15 + dx && j > 139 + dy)
+            {
+                set_pixel(self, i, j, pix);
+                printf("one: i:%d j:%d\n",i ,j);
+            }
+        }
+    }
+}
+
+void draw_zero(TGAImage *self, int dx, int dy, RGBA *pix){
+    for (int i = 0 + dx; i < 195 + dx; i++)
+    {
+        for (int j = 0 + dx; j < 169 + dy; j++)
+        {
+            if (i > 15 + dx && j > 139 + dy)
+            {
+                set_pixel(self, i, j, pix);
+                printf("one: i:%d j:%d\n",i ,j);
+            }
+        }
+    }
+}
+
+//nevim jestli tohle muze byt jen v mainu nebo pokuud pouziju typedef nejak na tohle pak to jde takhle
+//typedef of array of function pointers (takhle pka func_array je pak data type plus jmesno)
+//typedef TGAImage* (*func_array[])(TGAImage *, size_t, size_t, RGBA *);
+
+void watch_draw_time(TGAImage* self, const int hours, const int minutes){
```

Comment:
Nedava my smysl, ze `watch_draw_time` si neudela hlavicku samo, ale zapisuje do uz alokovaneho pole. Je to velice `unsafe` design.

Idealne chces mit "ridici" funkci, ktera rozdeli tu praci - to je podle me tato. Ocekavat ze uz budes mit nejake predpripravenou hlavicku... jakze dava to smysl pokud by jsi ocekaval ze to budes pokazde prekreslovat...

Pozor, ale musis tedy pokazde na zacatku "vynulovat" to pole, protoze jinak mas spoustu hodnot (ktere jsi nezakresloval) nedefinovane (a da ti to segfault pri cteni z te pameti, coz delas pri zapisu souboru)...

Reply:
tohle jsou cenny znalosti

to "pozor" myslis jako kdybych tu hlavicky vyrabel v te watch_draw jak navrhujes? Nebo tedka to nekde nenulujuu uz spravne?